### DISCUSSION TITLE
Incorporate custom optimisers?

### DISCUSSION THREAD

**Discussion by jpbreuer** [sherpa/sherpa#2042] [Ideas]
Hello! And to whom it may concern,

Nick from the help desk recommended I come here to start a discussion!

To keep it short, I've been playing around with some nature-inspired optimisers based on slime moulds, ant behaviour, etc. and have started to make some reasonable progress. For an example, here is a gif of a slime mould inspired algorithm finding the global minima given Rastrigin and Ackley functions:

![slime_mould_optimization_rastrigin](https://github.com/sherpa/sherpa/assets/11244366/605ebacc-ed33-4ab2-a9f5-63daefe6d557)

![slime_mould_optimization_ackley](https://github.com/sherpa/sherpa/assets/11244366/9901cabf-af75-480b-b62d-9647fec2e823)

It seems to work fine given multivariable landscapes and many local minima, but it starts to become confused when given a bimodal case with two equivalently appropriate solutions:

![slime_mould_optimization_bimodal1](https://github.com/sherpa/sherpa/assets/11244366/224e57d6-a1e4-45be-9cbd-72fdc636c8d7)

I am currently playing with some friction terms and other parameters that can help with convergence, I'm also looking at the parallelisation of agents and so forth. 

In any case, I am slowly coming to a conclusion that these methods are great when it comes to exploring and constraining a complex multidimensional parameter space, but perhaps are superseded by neldermead simplex and other better known methods when it comes to really converging on very precise global minima. My next task will be to merge this with other methods as a sort of post-processing to get the best of both worlds.

Anyways, I took a try at modifying my code to make it sherpa-compatible based on the scripts in ${ASCDS_LIB}/python3.11/site-packages/sherpa/optmethods , but I was wondering if any of the sherpa developers had on-hand some test data or test notebooks or something similar that I could use for evaluating the performance of the method within the sherpa environment or with sherpa tools. If not, then perhaps maybe I could get some suggestions of what I could try?

**Comment by hamogu** [comment]
Thanks for reaching out and especially sharing those cool plots! This is exciting to me on so many levels! For one, let me say that having more choices of optimizers in Sherpa has long been on our list (e.g. #580, #611, the unfinished #2005) to do because there simply is no single best optimizer that can work for any problem. Sure, some work generally well, but no optimizer works for everything.

In fact, two of the tabs in my webbrowser that are currently open point to [pypesto](https://pypesto.readthedocs.io/en/latest/example/getting_started.html) as an example for a package that makes very different optimizers from different packages available to its users, where I want to study if you could do the same for Sherpa. Another one is ESA's [pygmo2](https://pypesto.readthedocs.io/en/latest/example/getting_started.html) library, which also provides a bunch of very different optimizers (including an "artificial bee colony"), that I think might be useful to have available within Sherpa. 

So, I absolutely want to support your work in any way that can help!

## Test data for performance
What we don't currently have is what you are asking for: A comprehensive test suite that measures performance of optimizers. And to a certain degree, you can't have that because the best optimizer depends on your problem (see discussion in #1102). For example, if I fit the temperature of a stellar corona, I'm fine if I don't get to the absolute numerical minimum - my count numbers are typically very small and the systematics of the atomic data and the Chandra calibration are of order 10% anyway. On the other hand, it's important to me that my algorithm does not get stuck in a local minimum, because I often fit multiple temperature components with many different elemental abundances and sometimes with different absorption components. Now, in other situations that might be different, e.g. if you fit a powerlaw to a well-exposed AGN, then you should get much better than 10%, but maybe you are less worried about local minima, because you have fewer parameters.

Since most users of Sherpa are X-ray astronomers, I think we *should* have a notebook or something like that with three typical X-ray problems (and it's on my list of tasks for 2024 to write that), but there will always be problems that call for more specialized optimizers, which might not do well in those test problems. 

What I do now for looking at optimization, is to simply pick any X-ray dataset that I happen to have on my disk from the last science project I'm working on and then I refit that with different starting values and different settings and simply look at the results by eye.

## What I should do next to help you

I should write that test notebook. And I just learned about [pytest-harvest](https://smarie.github.io/python-pytest-harvest/) which looks like a great way to set that up. I won't get to it this week or next, but if that helps you, I can try to work on this very soon.

**Comment by jpbreuer** [comment]
That's great! I'm quite happy to hear that I came to the right place! I submitted my PhD thesis last week, so I am finally a bit more free to work on projects I've put on hold for the last years, including this one, but when I started talking about slime moulds during our last group meeting, I was met with much less enthusiasm. I was inspired to play around with this a bit after the astrostat talk on Repelling-Attracting Hamiltonian Monte Carlo from a week or two ago.

Anyways yes, in principle I agree with all your comments. I wanted to bring up making a test environment suite with test data and diagnostics because of these very reasons. For one, it would be nice to have some consensus about exactly what metrics and problems are necessary. I am a cxb+galaxy cluster person, so I am more biased that direction for example. I also work on low S/N outskirt regions and instrumental backgrounds with poor statistics, so I would like to have some robust example cases that incorporate more difficult-to-constrain problems. In the wake of XRISM and future calorimeter missions, having an example case with fitting the Fe/Ni-complex in the Perseus data would also be excellent. 

On that note, one important factor would be to document and standardise a framework to facilitate users to contribute or adapt their implementations into Sherpa. Would you require users to convert their code to C++? Would there be some limit to which libraries and dependencies would be acceptable? Making one big collaborative notebook that compares various optimisers would be excellent, and would be a great/interesting paper as well, potentially.

Leading on from that, it would be equally important to reach a consensus on how someone building the optimiser should approach and try to handle more complex problems, such as the bimodal case I posed above. E.g., when two equally appropriate solutions exist, would it be better to converge to either, neither, both, pick the middle point, return a 2d probability density map of the parameter space, switch to another optimiser, etc. 

I've dug through the scripts in the optmethods directory and the functions are quite well documented, but they are not necessarily suitable as a basis for someone to use them to build something beyond their scope, especially without being intimately familiar with Sherpa internals.

Please let me know how I can help! I would be happy to get involved on some development work, we would just need to come to a clear description of what is useful.

Edit: other things could also be more clear, for example regarding the convergence criteria and other things which are not documented explicitly.

**Reply by hamogu** [reply]
Sherpa is actually pretty flexible. Code in Sherpa itself can be in C++ or Python with Python preferred (but of course, numerics often need the speed of C++, which is why we do that); and there can be optional dependencies. For example, you can do the plotting either with matplotlib or with bokeh, or do no plots at all if neither is installed. You can read fits files either with crates or with astropy. Similarly, I envision that additional optimizers can have optional dependencies (e.g. if I add an optimizer that uses scipy and scipy is not installed, it will simply say "scipy not installed, please pick a different optimizer"). So if you write code that needs some other Python package that's fine; of course if it's all written from scratch by you than it can also also be all be part of Sherpa like the current optimizers.

We have a bunch of off-the-shelf C++ code (e.g. for FFT) that is compiled when you install Sherpa, but it would be nice to actually reduce that in the long-term (e.g. by pulling in other packages from pip or conda instead of shipping it with Sherpa itself) and not add to it.

**Reply by hamogu** [reply]
Convergence criteria etc. are up to you - every algorithm is different. It's easy to run a simple lev-mar to 1e-12, but that's just never going to work with a Monte-Carlo fitter in within the age of the universe; some parameters also don't make sense for all optimizers (e.g. "number of chains" only makes sense for MCMC). Thus, users can set that those as parameters for each optimizer they use, e.g. `xtol` and `rtol` in `LevMar` (https://sherpa.readthedocs.io/en/4.16.0/optimisers/api/sherpa.optmethods.LevMar.html). 
As we add more algorithms, it might makes sense to standardize that more and write is down, but as long as we have only a few optimizers that seemed to be more efforts than it's worth.  So, when you implement a new optimizer, I suggest reusing parameter names where it makes sense (such as `xtol`, `rtol`) for a standardized interface. Beyond that, we should see what happens and work out what's best when it comes up. If we need more standardized interfaces, I believe its easier to design them after you have a few examples then to work it all out before hand.

**Comment by hamogu** [comment]
As for test-data we have two places: For Sherpa alone, we have test data in  https://github.com/sherpa/sherpa-test-data
That's the test data we use for our CI tests. (The test suite will pass if you don't have it available, but will skip many tests, see https://sherpa.readthedocs.io/en/4.16.0/install.html#developer-build). 
We've also very recently (which means that there is not that much in there yet) started to collected test data here: https://github.com/HEACIT/curated-test-data with the intention of sharing that between different fitting packages (XSPEC, Sherpa, SPEX etc.) so we don't have to all use the same.
Of course, we can use any data for our performance testing, but it might just be convenient to re-use data that's already sitting in a repository so that anyone can clone that repro and then run the tests.

**Comment by hamogu** [comment]
There is, in fact, one PR open right now that does exactly what you suggest: Add another optimiser. I suggest to have a look: https://github.com/sherpa/sherpa/pull/2005

**Comment by DougBurke** [comment]
So, in #2022 I'm messing around with the internals of the fit/optimization/statistic code for reasons™, and as part of this trying to add approximate typing rules to make things a bit easier to understand - but the actual interface should be relatively "easy" (although looking at https://sherpa.readthedocs.io/en/4.16.0/optimisers/api/sherpa.optmethods.OptMethod.html it's not obvious, which is something to improve). You basically need an optimization function - e.g. we have

```python
lass LevMar(OptMethod):
    """Levenberg-Marquardt optimization method.
    ...
    """

    def __init__(self, name='levmar'):
        OptMethod.__init__(self, name, lmdif)

def lmdif(fcn, x0, xmin, xmax, ftol=EPSILON, xtol=EPSILON, gtol=EPSILON,
          maxfev=None, epsfcn=EPSILON, factor=100.0, numcores=1, verbose=0):
    ...
```

The important part is that it's roughly

```python
def optimizer(statfunc: Callable[[Sequence[float]], tuple(float, ndarray)],
                      x0: Sequence[float],  # or maybe ndarray[float]
                      xmin: Sequence[float],
                      xmax: Sequence[float]
                     ) -> tuple[bool, np.ndarray, float, str, dict]:
```

`statfunc` returns the statistic value and a per-bin statistic value for  the given set of parameters (unfortunately our optimisation expert left before we could get proper description of this return value), and `x0`/`xmin`/`xmax` are the starting parameter values and their limits (so they have the same size, and match the size of the sequence sent to `statfunc`. You'd have to go looking in `sherpa.optmethods.optfcts` to check exactly on the return values (or track prs like #2022, or create an issue), as we haven't concentrated on documenting this properly yet. 

One issue is that many optimizers do not include support for bounded parameters (ie our `xmin`/`xmax`  arguments). You can use something like an "infinite potential well" - that is, return <the biggest number you can think of> if a parameter value falls outside it's range - or something fancier. There is some code to do this in `optfcts` **but** at the moment it's not accessible, but maybe will become more accessible if I can get back to #2022.

**Comment by DougBurke** [comment]
I have just added

https://nbviewer.jupyter.org/github/DougBurke/sherpa-standalone-notebooks/blob/main/Optimization%20Example.ipynb

to my sherpa notebook collection at https://github.com/DougBurke/sherpa-standalone-notebooks?tab=readme-ov-file

I hope it's somewhat readable. Let me know if it is (or isn't) useful as I'd like to add something like this to the Sherpa documentation.

**Reply by jpbreuer** [reply]
Hi @DougBurke, I just went over your notebook and it's super cool! It's incredibly detailed, and your example not-an-optimiser case is quite intuitive to understand. The plots at the end are also very interesting; I'll be curious to make something similar with spectral fits to better understand how my own optimisers are performing. I won't be surprised if soon we see a bunch of people start writing their own optimisers :D

I guess the next pending task is to make the testing notebook with different tests and example cases that @hamogu and I were discussing so we can trully validate the performance of the different methods.

In any case, thanks a lot for this notebook, it's really quite awesome!

**Comment by hamogu** [comment]
I've added a basic framework for benchmarking to https://github.com/hamogu/sherpa-opt-benchmark (we might move that repro from my personal GH account to sherpa at some point in the future).
The goal is not to benchmark each and every possible dataset, but rather to set up a framework that can easily be edited to throw in another optimizer and see how it performs compared to the built-in optimizers.

Currently, I have two different X-ray daatsets as example (one used twice with different starting values for the fit); that could obviously be expanded. I've decided to start with "real" problems (OK, simulated data, but real ARFs, RMFs, and spectral models that I might actually use), because that's what Sherpa is mostly used for. I envision adding a few more of those. In terms of benchmarking performance, I think that's more relevant for us than the artificial test problems that are used in the academic literature to evaluate how fitters are doing. 
In Sherpa, we know something about our data - most users fit spectra with global models or part of spectra with emission or absorption lines - so I think that's the place to start. We'll have to see how we use this in practice and add data and setups (maybe with parameters boundaries) when we need it.

**Reply by DougBurke** [reply]
For background, Sherpa does have a bunch of tests that it uses to check optimizers out (it depends on what you mean by benchmark here - is it "how fast" or "does it converge" or "how fast does it converge"). Unfortunately Dan didn't provide much information on how he used the code - there is #810 and #849 and then I did #1778

In https://github.com/sherpa/sherpa/blob/main/sherpa/optmethods/tests/test_optmethods.py we use some of the tests from 

J. MORE', B. GARBOW & K. HILLSTROM, "Algorithm 566: Fortran Subroutines for Testing Unconstrained Optimization Software.", ACM TOMS, VOL. 7, PAGES 14-41 AND 136-140, 1981

or maybe 

More, J.J., Garbow, B.S. and Hillstrom, K.E., Testing Unconstrained Optimization Software, ACM Trans. Math. Software 7 (1981), 17-41.

(which I think is essentially the above but with a different page number) or maybe

More, J. J., Garbow, B. S., and Hillstrom, K. E.. Testing unconstrained optimization software. United States: N. p., 1978. Web. doi:10.2172/6650344.

which we can get from https://doi.org/10.2172/6650344

but these tests are always tricky to understand (since they are not set up with the data/model approach Sherpa uses, and then the code has some way of repeating the values to be tested)

There is also - from #1778

https://github.com/sherpa/sherpa/blob/main/sherpa/optmethods/tests/test_opt_original.py

which can be run directly (rather than via `pytest`) but I really don't understand how Dan used this output to check on the optimisers.

I will note that test_optmethods.py is not ideal

- it should test at the object rather than UI level
- it should not need so many tests that require XSPEC

but this is not an area I'm likely to work on at any time soon.

**Reply by hamogu** [reply]
I meant that file `test_optmethods.py` (which I looked at before I wrote this) when I said "artificial test problems that are used in the academic literature". It's good to have those - but it's complimentary to benchmarking on X-ray datasets with XSPEC models, which is how most of our users use Sherpa.

**Comment by DougBurke** [comment]
Just to note that the [4.18.0 release](https://zenodo.org/records/17288983) will now allow you to use some of the optimizers from SciPy or optimagic if installed. The documentation has also been improved around this area (of adding optimizers). See #2303 for more details .

**Reply by jpbreuer** [reply]
Exciting! Thank you for your effort! :D
